<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Выгрузка-и-анализ-данных" data-toc-modified-id="Выгрузка-и-анализ-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Выгрузка и анализ данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Удаление-не-актуальных-столбцов" data-toc-modified-id="Удаление-не-актуальных-столбцов-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Удаление не актуальных столбцов</a></span></li></ul></li></ul></li><li><span><a href="#Обучение-моделей" data-toc-modified-id="Обучение-моделей-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение моделей</a></span><ul class="toc-item"><li><span><a href="#REV02-Альтернативный-способ-подготовки-данных-и-обучения-модели" data-toc-modified-id="REV02-Альтернативный-способ-подготовки-данных-и-обучения-модели-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>REV02 Альтернативный способ подготовки данных и обучения модели</a></span></li><li><span><a href="#Преобразование-модели" data-toc-modified-id="Преобразование-модели-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Преобразование модели</a></span></li><li><span><a href="#Подготовка-обучающей-выборки" data-toc-modified-id="Подготовка-обучающей-выборки-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Подготовка обучающей выборки</a></span></li><li><span><a href="#Обучение-моделей" data-toc-modified-id="Обучение-моделей-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Обучение моделей</a></span><ul class="toc-item"><li><span><a href="#Linear-Regression" data-toc-modified-id="Linear-Regression-2.4.1"><span class="toc-item-num">2.4.1&nbsp;&nbsp;</span>Linear Regression</a></span></li><li><span><a href="#Ridge" data-toc-modified-id="Ridge-2.4.2"><span class="toc-item-num">2.4.2&nbsp;&nbsp;</span>Ridge</a></span></li><li><span><a href="#DecisionTreeRegresso" data-toc-modified-id="DecisionTreeRegresso-2.4.3"><span class="toc-item-num">2.4.3&nbsp;&nbsp;</span>DecisionTreeRegresso</a></span></li><li><span><a href="#CatBoostRegressor" data-toc-modified-id="CatBoostRegressor-2.4.4"><span class="toc-item-num">2.4.4&nbsp;&nbsp;</span>CatBoostRegressor</a></span></li><li><span><a href="#Случайный-Лес" data-toc-modified-id="Случайный-Лес-2.4.5"><span class="toc-item-num">2.4.5&nbsp;&nbsp;</span>Случайный Лес</a></span></li><li><span><a href="#LGBMRegressor" data-toc-modified-id="LGBMRegressor-2.4.6"><span class="toc-item-num">2.4.6&nbsp;&nbsp;</span>LGBMRegressor</a></span></li></ul></li></ul></li><li><span><a href="#Анализ-моделей" data-toc-modified-id="Анализ-моделей-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Анализ моделей</a></span><ul class="toc-item"><li><span><a href="#Объединение-и-анализ-полученных-результатов" data-toc-modified-id="Объединение-и-анализ-полученных-результатов-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Объединение и анализ полученных результатов</a></span></li><li><span><a href="#Проверка-финальной-метрики" data-toc-modified-id="Проверка-финальной-метрики-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Проверка финальной метрики</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Определение стоимости автомобилей

Сервис по продаже автомобилей с пробегом «Не бит, не крашен» разрабатывает приложение для привлечения новых клиентов. В нём можно быстро узнать рыночную стоимость своего автомобиля. В вашем распоряжении исторические данные: технические характеристики, комплектации и цены автомобилей. Вам нужно построить модель для определения стоимости. 

Заказчику важны:

- качество предсказания;
- скорость предсказания;
- время обучения.

**Заказчик** — Сервис по продаже автомобилей с пробегом «Не бит, не крашен».

**Цель исследования** 

* Построение модели для предсказания цены на поддержанный автомобиль

**Входные данные**  — ```/datasets/autos.csv``` - база данных пользователей


**Результаты исследования** — Получена модель подходящая под требования заказчика

## Подготовка данных

### Выгрузка и анализ данных

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score


from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge

from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error

from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import StandardScaler

from sklearn.tree import DecisionTreeRegressor

#!pip install catboost
from catboost import Pool
from catboost import CatBoostRegressor
from catboost import cv

#!pip install lightgbm
from lightgbm import LGBMRegressor

In [2]:
# выгрузим данные и посмотрим в каком виде получается датафрейм
df = pd.read_csv('datasets/autos.csv')
df.rename(columns = {'Repaired':'NotRepaired'}, inplace = True )
df.head()


DateCrawled  Price VehicleType  RegistrationYear Gearbox  Power  \
0  2016-03-24 11:52:17    480         NaN              1993  manual      0   
1  2016-03-24 10:58:45  18300       coupe              2011  manual    190   
2  2016-03-14 12:52:21   9800         suv              2004    auto    163   
3  2016-03-17 16:54:04   1500       small              2001  manual     75   
4  2016-03-31 17:25:20   3600       small              2008  manual     69   

   Model  Kilometer  RegistrationMonth  FuelType       Brand NotRepaired  \
0   golf     150000                  0    petrol  volkswagen         NaN   
1    NaN     125000                  5  gasoline        audi         yes   
2  grand     125000                  8  gasoline        jeep         NaN   
3   golf     150000                  6    petrol  volkswagen          no   
4  fabia      90000                  7  gasoline       skoda          no   

           DateCreated  NumberOfPictures  PostalCode             LastSeen  
0  2016-03-24 00:00:00                 0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00                 0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00                 0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00                 0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00                 0       60437  2016-04-06 10:17:21

**Описание таблицы:**

Признаки
* `DateCrawled` — дата скачивания анкеты из базы
* `VehicleType` — тип автомобильного кузова
* `RegistrationYear` — год регистрации автомобиля
* `Gearbox` — тип коробки передач
* `Power` — мощность (л. с.)
* `Model` — модель автомобиля
* `Kilometer` — пробег (км)
* `RegistrationMonth` — месяц регистрации автомобиля
* `FuelType` — тип топлива
* `Brand` — марка автомобиля
* `NotRepaired` — была машина в ремонте или нет (скорее должно быть не ремонтировалась)
* `DateCreated` — дата создания анкеты
* `NumberOfPictures` — количество фотографий автомобиля
* `PostalCode` — почтовый индекс владельца анкеты (пользователя)
* `LastSeen` — дата последней активности пользователя

Целевой признак
* `Price` — цена (евро)

In [3]:
# изучим датафрейм
# функция для обработки данных датафрейма
def research(df):
    
    display(df)
    print('Информация по таблице:', '\n')
    df.info()
    print()
    print('Количество пропусков:', '\n')
    print(df.isna().sum(), '\n', '\n')
    print('Процент пропусков:', '\n')
    print(df.isna().mean()*100, '\n')

    display(df.describe())
    print()
    print('Число явных дубликатов:', df.duplicated().sum(), '\n')


In [4]:
research(df)

DateCrawled  Price  VehicleType  RegistrationYear Gearbox  \
0       2016-03-24 11:52:17    480          NaN              1993  manual   
1       2016-03-24 10:58:45  18300        coupe              2011  manual   
2       2016-03-14 12:52:21   9800          suv              2004    auto   
3       2016-03-17 16:54:04   1500        small              2001  manual   
4       2016-03-31 17:25:20   3600        small              2008  manual   
...                     ...    ...          ...               ...     ...   
354364  2016-03-21 09:50:58      0          NaN              2005  manual   
354365  2016-03-14 17:48:27   2200          NaN              2005     NaN   
354366  2016-03-05 19:56:21   1199  convertible              2000    auto   
354367  2016-03-19 18:57:12   9200          bus              1996  manual   
354368  2016-03-20 19:41:08   3400        wagon              2002  manual   

        Power        Model  Kilometer  RegistrationMonth  FuelType  \
0           0         golf     150000                  0    petrol   
1         190          NaN     125000                  5  gasoline   
2         163        grand     125000                  8  gasoline   
3          75         golf     150000                  6    petrol   
4          69        fabia      90000                  7  gasoline   
...       ...          ...        ...                ...       ...   
354364      0         colt     150000                  7    petrol   
354365      0          NaN      20000                  1       NaN   
354366    101       fortwo     125000                  3    petrol   
354367    102  transporter     150000                  3  gasoline   
354368    100         golf     150000                  6  gasoline   

                 Brand NotRepaired          DateCreated  NumberOfPictures  \
0           volkswagen         NaN  2016-03-24 00:00:00                 0   
1                 audi         yes  2016-03-24 00:00:00                 0   
2                 jeep         NaN  2016-03-14 00:00:00                 0   
3           volkswagen          no  2016-03-17 00:00:00                 0   
4                skoda          no  2016-03-31 00:00:00                 0   
...                ...         ...                  ...               ...   
354364      mitsubishi         yes  2016-03-21 00:00:00                 0   
354365  sonstige_autos         NaN  2016-03-14 00:00:00                 0   
354366           smart          no  2016-03-05 00:00:00                 0   
354367      volkswagen          no  2016-03-19 00:00:00                 0   
354368      volkswagen         NaN  2016-03-20 00:00:00                 0   

        PostalCode             LastSeen  
0            70435  2016-04-07 03:16:57  
1            66954  2016-04-07 01:46:50  
2            90480  2016-04-05 12:47:46  
3            91074  2016-03-17 17:40:17  
4            60437  2016-04-06 10:17:21  
...            ...                  ...  
354364        2694  2016-03-21 10:42:49  
354365       39576  2016-04-06 00:46:52  
354366       26135  2016-03-11 18:17:12  
354367       87439  2016-04-07 07:15:26  
354368       40764  2016-03-24 12:45:21  

[354369 rows x 16 columns]

Информация по таблице: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   DateCrawled        354369 non-null  object
 1   Price              354369 non-null  int64 
 2   VehicleType        316879 non-null  object
 3   RegistrationYear   354369 non-null  int64 
 4   Gearbox            334536 non-null  object
 5   Power              354369 non-null  int64 
 6   Model              334664 non-null  object
 7   Kilometer          354369 non-null  int64 
 8   RegistrationMonth  354369 non-null  int64 
 9   FuelType           321474 non-null  object
 10  Brand              354369 non-null  object
 11  NotRepaired        283215 non-null  object
 12  DateCreated        354369 non-null  object
 13  NumberOfPictures   354369 non-null  int64 
 14  PostalCode         354369 non-null  int64 
 15  LastSeen           354369 non-null  object


Price  RegistrationYear          Power      Kilometer  \
count  354369.000000     354369.000000  354369.000000  354369.000000   
mean     4416.656776       2004.234448     110.094337  128211.172535   
std      4514.158514         90.227958     189.850405   37905.341530   
min         0.000000       1000.000000       0.000000    5000.000000   
25%      1050.000000       1999.000000      69.000000  125000.000000   
50%      2700.000000       2003.000000     105.000000  150000.000000   
75%      6400.000000       2008.000000     143.000000  150000.000000   
max     20000.000000       9999.000000   20000.000000  150000.000000   

       RegistrationMonth  NumberOfPictures     PostalCode  
count      354369.000000          354369.0  354369.000000  
mean            5.714645               0.0   50508.689087  
std             3.726421               0.0   25783.096248  
min             0.000000               0.0    1067.000000  
25%             3.000000               0.0   30165.000000  
50%             6.000000               0.0   49413.000000  
75%             9.000000               0.0   71083.000000  
max            12.000000               0.0   99998.000000


Число явных дубликатов: 4 



**Промежуточный итог:**
* Наименование столбцов выполнены в CamelCase, что в целом удовлетворяет нормам
* Тип столбцов: *(но можно не менять так как в дальнейшем будут удалены)*
    * `DateCrawled` - object -> date
    * `DateCreated` - object -> date (можно отбросить часы так как везде все равно нули)
    * `LastSeen` - object -> date
* Пропуски: (глянуть общие пропуски. Изучить по их кретичности, могут быть зависимы)
    * `NotRepaired` - 20%
    * `VehicleType` - 11%
    * `FuelType` - 9%
    * `Model` и `Gearbox` - 5%
* Аномальные данные:
    * `RegistrationYear` - 9999
    * `Power` - 20000 (походу истребитель продают)
    * `Price` - 0 *(бесплатно, это я люблю, беру)*
* Имеется 4 явных дубликата - можно удалить, так как есть идентификатор пользователя `PostalCode`, а значит скорее всего создал 4 одинаковых объявления (можно глянуть)
* Имеются признаки, которые можно удалить в конце обработки так как слабо влияют на реальную цену автомобиля:
    * `PostalCode` - кто продает не так важно
    * `DateCrawled`, `DateCreated`, `LastSeen` - даты могут быть важны при увелечении цены из-за различных событий связанных с датой. Например введение новой таможенной пошлины, различных ограничений или ухода компании с рынка. 
    * `NumberOfPictures` - количество фото, не так сильно влияют на саму стоиомсть авто, но могут использоваться для обучения других моделей... ad-hoc
    



**Необходима предобработка данных**


### Предобработка данных

*- Изучите данные. Заполните пропущенные значения и обработайте аномалии в столбцах. Если среди признаков имеются неинформативные, удалите их.*

In [5]:
# удаление явных дубликатов 
df = df.drop_duplicates().reset_index(drop=True) 
print('Число явных дубликатов:', df.duplicated().sum(), '\n')

Число явных дубликатов: 0 



In [6]:
# посмотрим на данные в ячейках, нет ли разного написания и т.д.

print('VehicleType ->', df['VehicleType'].sort_values().unique(), '\n')
print('Gearbox ->', df['Gearbox'].sort_values().unique(), '\n')

# особо не читаемо, много моделей, не сортируются 
print('Model ->', df['Model'].sort_values().unique(), '\n') 

print('FuelType ->', df['FuelType'].sort_values().unique(), '\n') 

print('Brand ->', df['Brand'].sort_values().unique(), '\n') 

print('NotRepaired ->', df['NotRepaired'].sort_values().unique(), '\n') 

VehicleType -> ['bus' 'convertible' 'coupe' 'other' 'sedan' 'small' 'suv' 'wagon' nan] 

Gearbox -> ['auto' 'manual' nan] 

Model -> ['100' '145' '147' '156' '159' '1_reihe' '1er' '200' '2_reihe' '300c'
 '3_reihe' '3er' '4_reihe' '500' '5_reihe' '5er' '601' '6_reihe' '6er'
 '7er' '80' '850' '90' '900' '9000' '911' 'a1' 'a2' 'a3' 'a4' 'a5' 'a6'
 'a8' 'a_klasse' 'accord' 'agila' 'alhambra' 'almera' 'altea' 'amarok'
 'antara' 'arosa' 'astra' 'auris' 'avensis' 'aveo' 'aygo' 'b_klasse'
 'b_max' 'beetle' 'berlingo' 'bora' 'boxster' 'bravo' 'c1' 'c2' 'c3' 'c4'
 'c5' 'c_klasse' 'c_max' 'c_reihe' 'caddy' 'calibra' 'captiva' 'carisma'
 'carnival' 'cayenne' 'cc' 'ceed' 'charade' 'cherokee' 'citigo' 'civic'
 'cl' 'clio' 'clk' 'clubman' 'colt' 'combo' 'cooper' 'cordoba' 'corolla'
 'corsa' 'cr_reihe' 'croma' 'crossfire' 'cuore' 'cx_reihe' 'defender'
 'delta' 'discovery' 'doblo' 'ducato' 'duster' 'e_klasse' 'elefantino'
 'eos' 'escort' 'espace' 'exeo' 'fabia' 'fiesta' 'focus' 'forester'
 'forfour' 'f

####  Обработка пропусков

Начнем с `NotRepaired` (не ремонтировалась), около 20% пропусков, просто удалить не можем, значит надо думать. Возможно пользователи которые не чинили, просто не поставили в выборе наличие ремонта (Нет). По этой причине кажется рациональным поставить вместо пропусков "да"

In [7]:
df["NotRepaired"].fillna("yes", inplace=True)

In [8]:
# Посмотрим, что там с аномальными датами возьмем дату создания первого автомобиля 1886 г. и текущей даты 2022 г.
print("Колличество объявлений с некорректной годом:", len(
    df.loc[(df['RegistrationYear'] > 2016) 
           | (df['RegistrationYear'] < 1886) 
           | (df['RegistrationYear'] == 2016) & (df['RegistrationMonth'] > 9)]))

# год производства лучше всего взять по группе Brand + Model + VehicleType и по хорошему сгруппировать по среднему пробегу

df = df.loc[(df['RegistrationYear'] <= 2016) & (df['RegistrationYear'] >= 1886)]

Колличество объявлений с некорректной годом: 16177


Посмотрим, что у нас по мощности, возьмем за ограничение Lotus Evija (самый мощный серийный автомобиль в мире) — 2000 л.с.
Так же можно глянуть сколько продают гиперкаров - от 1000 до 2000 л.с

In [9]:
print("Колличество объявлений с некорректной мощностью:", len(df.loc[(df['Power'] > 2000) | (df['Power'] <= 0)]))
print("Колличество гиперкаров:", len(df.loc[(df['Power'] >= 1000) & (df['Power'] <= 2000)]))



Колличество объявлений с некорректной мощностью: 36562
Колличество гиперкаров: 172


In [10]:
# Посмотрим, что там с нулевой стоимостью


print('Количество машин с нулевой ценой:', len(df.loc[df['Price'] == 0]))
print('Процент машин с нулевой ценой:', len(df.loc[df['Price'] == 0])/len(df)*100, '\n')



display(df.loc[df['Price'] == 0])

Количество машин с нулевой ценой: 10008
Процент машин с нулевой ценой: 2.945530639934779 



DateCrawled  Price VehicleType  RegistrationYear Gearbox  \
7       2016-03-21 18:54:38      0       sedan              1980  manual   
40      2016-03-26 22:06:17      0         NaN              1990     NaN   
115     2016-03-20 18:53:27      0       small              1999     NaN   
152     2016-03-11 18:55:53      0         bus              2004  manual   
154     2016-03-22 16:37:59      0         NaN              2006     NaN   
...                     ...    ...         ...               ...     ...   
354201  2016-03-09 15:56:30      0         NaN              2000  manual   
354234  2016-03-20 14:55:07      0       small              2002  manual   
354244  2016-03-24 13:48:05      0       small              1999  manual   
354273  2016-03-10 22:55:50      0       small              1999  manual   
354360  2016-03-21 09:50:58      0         NaN              2005  manual   

        Power   Model  Kilometer  RegistrationMonth FuelType       Brand  \
7          50   other      40000                  7   petrol  volkswagen   
40          0   corsa     150000                  1   petrol        opel   
115         0     NaN       5000                  0   petrol  volkswagen   
152       101  meriva     150000                 10      lpg        opel   
154         0   other       5000                  0      NaN        fiat   
...       ...     ...        ...                ...      ...         ...   
354201     65   corsa     150000                  0      NaN        opel   
354234     60  fiesta     150000                  3   petrol        ford   
354244     53   swift     150000                  3   petrol      suzuki   
354273     37   arosa     150000                  7   petrol        seat   
354360      0    colt     150000                  7   petrol  mitsubishi   

       NotRepaired          DateCreated  NumberOfPictures  PostalCode  \
7               no  2016-03-21 00:00:00                 0       19348   
40             yes  2016-03-26 00:00:00                 0       56412   
115            yes  2016-03-20 00:00:00                 0       37520   
152            yes  2016-03-11 00:00:00                 0       27432   
154            yes  2016-03-22 00:00:00                 0       44867   
...            ...                  ...               ...         ...   
354201         yes  2016-03-09 00:00:00                 0       23758   
354234         yes  2016-03-20 00:00:00                 0       33659   
354244         yes  2016-03-24 00:00:00                 0       42329   
354273         yes  2016-03-10 00:00:00                 0       22559   
354360         yes  2016-03-21 00:00:00                 0        2694   

                   LastSeen  
7       2016-03-25 16:47:58  
40      2016-03-27 17:43:34  
115     2016-04-07 02:45:22  
152     2016-03-12 23:47:10  
154     2016-04-06 06:17:56  
...                     ...  
354201  2016-03-30 11:16:08  
354234  2016-04-06 18:45:23  
354244  2016-04-07 05:17:24  
354273  2016-03-12 23:46:32  
354360  2016-03-21 10:42:49  

[10008 rows x 16 columns]

Можем ли мы угадывать или менять на медианное значение `Price` (цену автомобиля), что является Целевым признаком по которому будет сверяться модель? Мне кажется, что нет. 
Поэтому оставим в датафрейме только значения не равные 0

In [11]:
df = df.loc[df['Price'] != 0]

print('Количество машин с нулевой ценой:', len(df.loc[df['Price'] == 0]))
print('Процент машин с нулевой ценой:', len(df.loc[df['Price'] == 0])/len(df)*100, '\n')



Количество машин с нулевой ценой: 0
Процент машин с нулевой ценой: 0.0 



In [12]:
display(df[df['Model'].isna()])

DateCrawled  Price VehicleType  RegistrationYear Gearbox  \
1       2016-03-24 10:58:45  18300       coupe              2011  manual   
59      2016-03-29 15:48:15      1         suv              1994  manual   
81      2016-04-03 12:56:45    350       small              1997  manual   
135     2016-03-27 20:51:23   1450       sedan              1992  manual   
151     2016-03-27 20:47:22   6799       small              2009     NaN   
...                     ...    ...         ...               ...     ...   
354241  2016-03-07 16:37:42    560       small              2001    auto   
354317  2016-03-15 13:52:34   9400       wagon              2007  manual   
354334  2016-03-31 19:52:33    180         NaN              1995     NaN   
354347  2016-03-11 23:40:32   1900         NaN              2000  manual   
354361  2016-03-14 17:48:27   2200         NaN              2005     NaN   

        Power Model  Kilometer  RegistrationMonth  FuelType           Brand  \
1         190   NaN     125000                  5  gasoline            audi   
59        286   NaN     150000                 11       NaN  sonstige_autos   
81         54   NaN     150000                  3       NaN            fiat   
135       136   NaN     150000                  0       NaN            audi   
151        60   NaN      20000                  5    petrol      volkswagen   
...       ...   ...        ...                ...       ...             ...   
354241    170   NaN      90000                  0    petrol            fiat   
354317    200   NaN     150000                  4  gasoline  sonstige_autos   
354334      0   NaN     125000                  3    petrol            opel   
354347    110   NaN     150000                  7       NaN      volkswagen   
354361      0   NaN      20000                  1       NaN  sonstige_autos   

       NotRepaired          DateCreated  NumberOfPictures  PostalCode  \
1              yes  2016-03-24 00:00:00                 0       66954   
59             yes  2016-03-29 00:00:00                 0       53721   
81             yes  2016-04-03 00:00:00                 0       45665   
135             no  2016-03-27 00:00:00                 0       38709   
151             no  2016-03-27 00:00:00                 0       89077   
...            ...                  ...               ...         ...   
354241         yes  2016-03-07 00:00:00                 0       55743   
354317         yes  2016-03-15 00:00:00                 0       91186   
354334         yes  2016-03-31 00:00:00                 0       41470   
354347          no  2016-03-11 00:00:00                 0       87700   
354361         yes  2016-03-14 00:00:00                 0       39576   

                   LastSeen  
1       2016-04-07 01:46:50  
59      2016-04-06 01:44:38  
81      2016-04-05 11:47:13  
135     2016-04-05 20:17:31  
151     2016-03-27 20:47:22  
...                     ...  
354241  2016-03-12 21:45:25  
354317  2016-03-17 08:16:25  
354334  2016-04-06 14:18:04  
354347  2016-03-12 14:16:51  
354361  2016-04-06 00:46:52  

[15637 rows x 16 columns]

In [13]:
(
df[df['Model'].isna()]
 .pivot_table(index=['Brand', 'VehicleType'], columns='RegistrationYear', values='Power', aggfunc= 'count', fill_value='X')
)

RegistrationYear       1910 1925 1929 1934 1935 1937 1938 1940 1941 1943  ...  \
Brand      VehicleType                                                    ...   
alfa_romeo coupe          X    X    X    X    X    X    X    X    X    X  ...   
           other          X    X    X    X    X    X    X    X    X    X  ...   
           sedan          X    X    X    X    X    X    X    X    X    X  ...   
           small          X    X    X    X    X    X    X    X    X    X  ...   
           wagon          X    X    X    X    X    X    X    X    X    X  ...   
...                     ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
volkswagen suv            X    X    X    X    X    X    X    X    X    X  ...   
           wagon          X    X    X    X    X    X    X    X    X    X  ...   
volvo      coupe          X    X    X    X    X    X    X    X    X    X  ...   
           sedan          X    X    X    X    X    X    X    X    X    X  ...   
           wagon          X    X    X    X    X    X    X    X    X    X  ...   

RegistrationYear        2007 2008 2009 2010 2011 2012 2013 2014 2015 2016  
Brand      VehicleType                                                     
alfa_romeo coupe         1.0    X    X    X    X    X    X    X    X    X  
           other           X    X    X    X  1.0    X    X    X    X    X  
           sedan         1.0    X    X  1.0    X    X    X    X    X  1.0  
           small         1.0    X    X    X    X    X    X    X    X    X  
           wagon           X  1.0    X    X    X    X    X    X    X    X  
...                      ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  
volkswagen suv             X  1.0  3.0  2.0  1.0  1.0    X    X    X    X  
           wagon        12.0  9.0  5.0  7.0  2.0  4.0  2.0  2.0    X    X  
volvo      coupe           X    X    X    X    X    X    X    X    X    X  
           sedan         1.0    X    X    X    X  2.0    X    X    X    X  
           wagon           X    X    X    X    X    X    X    X    X    X  

[252 rows x 81 columns]

С моделью можно поступить одним из трех способов:
* вычеслить по группе схожих параметров: год выпуска, бренд, тип кузова и тип топлива
* поставить `unknown`
* удалить

In [14]:
# так как данных около 5% удалять не хочется, предлагаю поставить на unknown

In [15]:
# Исправим остальное:

df["FuelType"].fillna("other", inplace=True)

df["VehicleType"].fillna("other", inplace=True)

#заполнять Gearbox (коробка передачь) по средней или иной связке не корректно, 
#так как одни и теж модели, одного года, имеют разные типы в зависимости от комплектации (стоимости)
df["Gearbox"].fillna("unknown", inplace=True)


df["Model"].fillna("unknown", inplace=True)


# общий процент пропусков посмотрим
print('Процент пропусков:', '\n')
print(df.isna().mean()*100, '\n')

Процент пропусков: 

DateCrawled          0.0
Price                0.0
VehicleType          0.0
RegistrationYear     0.0
Gearbox              0.0
Power                0.0
Model                0.0
Kilometer            0.0
RegistrationMonth    0.0
FuelType             0.0
Brand                0.0
NotRepaired          0.0
DateCreated          0.0
NumberOfPictures     0.0
PostalCode           0.0
LastSeen             0.0
dtype: float64 



In [16]:
# обработаем данные с мощностью
df.loc[(df['Power'] > 2000) | (df['Power'] <= 0), 'Power'] = None
df['Power'] = df['Power'].fillna(df.groupby('Model')['Power'].transform('median'))


print("Колличество объявлений с некорректной мощностью:", len(df.loc[(df['Power'] > 2000) | (df['Power'] <= 0)]), '\n')

print(df['Power'])
print(df.isna().sum(), '\n')
# посмотрим, что тут не заполнилось
display(df[df['Power'].isna()])
display(df[df['Model'] == "serie_1"])
#такая только одна модель, удалим

df = df.loc[~df['Power'].isna()]


df['Power'] = df['Power'].astype('int64')
print()
print(df.info())


Колличество объявлений с некорректной мощностью: 0 

0         101.0
1         190.0
2         163.0
3          75.0
4          69.0
          ...  
354359    125.0
354361    106.0
354362    101.0
354363    102.0
354364    100.0
Name: Power, Length: 329761, dtype: float64
DateCrawled          0
Price                0
VehicleType          0
RegistrationYear     0
Gearbox              0
Power                1
Model                0
Kilometer            0
RegistrationMonth    0
FuelType             0
Brand                0
NotRepaired          0
DateCreated          0
NumberOfPictures     0
PostalCode           0
LastSeen             0
dtype: int64 



DateCrawled  Price VehicleType  RegistrationYear Gearbox  \
234294  2016-03-30 11:39:08   3800       wagon              1978  manual   

        Power    Model  Kilometer  RegistrationMonth  FuelType       Brand  \
234294    NaN  serie_1      30000                  0  gasoline  land_rover   

       NotRepaired          DateCreated  NumberOfPictures  PostalCode  \
234294         yes  2016-03-30 00:00:00                 0       49824   

                   LastSeen  
234294  2016-03-30 11:39:08

DateCrawled  Price VehicleType  RegistrationYear Gearbox  \
234294  2016-03-30 11:39:08   3800       wagon              1978  manual   

        Power    Model  Kilometer  RegistrationMonth  FuelType       Brand  \
234294    NaN  serie_1      30000                  0  gasoline  land_rover   

       NotRepaired          DateCreated  NumberOfPictures  PostalCode  \
234294         yes  2016-03-30 00:00:00                 0       49824   

                   LastSeen  
234294  2016-03-30 11:39:08


<class 'pandas.core.frame.DataFrame'>
Int64Index: 329760 entries, 0 to 354364
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   DateCrawled        329760 non-null  object
 1   Price              329760 non-null  int64 
 2   VehicleType        329760 non-null  object
 3   RegistrationYear   329760 non-null  int64 
 4   Gearbox            329760 non-null  object
 5   Power              329760 non-null  int64 
 6   Model              329760 non-null  object
 7   Kilometer          329760 non-null  int64 
 8   RegistrationMonth  329760 non-null  int64 
 9   FuelType           329760 non-null  object
 10  Brand              329760 non-null  object
 11  NotRepaired        329760 non-null  object
 12  DateCreated        329760 non-null  object
 13  NumberOfPictures   329760 non-null  int64 
 14  PostalCode         329760 non-null  int64 
 15  LastSeen           329760 non-null  object
dtypes: int64(7), object

#### Удаление не актуальных столбцов

In [17]:
end_df = df.drop(['DateCrawled', 'DateCreated', 'LastSeen', 
                  'NumberOfPictures', 'PostalCode', 'RegistrationMonth'], axis = 1)
end_df = end_df.reset_index(drop=True)

In [18]:
end_df.info()
end_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 329760 entries, 0 to 329759
Data columns (total 10 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   Price             329760 non-null  int64 
 1   VehicleType       329760 non-null  object
 2   RegistrationYear  329760 non-null  int64 
 3   Gearbox           329760 non-null  object
 4   Power             329760 non-null  int64 
 5   Model             329760 non-null  object
 6   Kilometer         329760 non-null  int64 
 7   FuelType          329760 non-null  object
 8   Brand             329760 non-null  object
 9   NotRepaired       329760 non-null  object
dtypes: int64(4), object(6)
memory usage: 25.2+ MB


Price  RegistrationYear          Power      Kilometer
count  329760.000000     329760.000000  329760.000000  329760.000000
mean     4607.036096       2002.593674     120.046843  128206.316715
std      4546.267728          6.927782      61.371297   37524.363197
min         1.000000       1910.000000       1.000000    5000.000000
25%      1200.000000       1999.000000      75.000000  125000.000000
50%      2900.000000       2003.000000     110.000000  150000.000000
75%      6600.000000       2007.000000     145.000000  150000.000000
max     20000.000000       2016.000000    2000.000000  150000.000000

**Промежуточные итоги**

* удалены явные дубликаты
* проблем в названии столбцов нет
* столбцы не актуальные для тестирования модели, удалены
* были подчищены, удалены или обновлены пропуски и аномальные данные

## Обучение моделей

* *Подготовьте выборки для обучения моделей.*
* *Обучите разные модели, одна из которых — LightGBM, как минимум одна — не бустинг. Для каждой модели попробуйте разные гиперпараметры.*


* *Для оценки качества моделей применяйте метрику RMSE.*
* *Значение метрики RMSE должно быть меньше 2500.*

In [19]:
df_test = end_df

In [20]:
rev02_df = end_df

### REV02 Альтернативный способ подготовки данных и обучения модели

In [21]:
rev02_target = rev02_df['Price']
rev02_features = rev02_df.drop('Price', axis=1)


In [22]:
rev02_features_train, rev02_features_test, rev02_target_train, rev02_target_test = train_test_split(
    rev02_features, rev02_target, test_size=0.25, random_state=12345)

In [23]:
print(rev02_features_train.shape, '\n')
display(rev02_features_train)

(247320, 9) 



VehicleType  RegistrationYear Gearbox  Power     Model  Kilometer  \
92955        coupe              1999  manual    107    megane     150000   
8833         small              2005  manual    140        a3     150000   
124632       sedan              2005  manual     75      golf     100000   
233625       wagon              1997  manual    184       5er     150000   
272265       wagon              2007  manual    143   6_reihe     150000   
...            ...               ...     ...    ...       ...        ...   
158838       wagon              1995    auto    150  e_klasse     150000   
47873        sedan              1991  manual     54      golf     150000   
86398        wagon              2011  manual    111   3_reihe      90000   
77285        wagon              2000  manual    101    passat     125000   
217570       wagon              2006  manual    131    megane     150000   

        FuelType          Brand NotRepaired  
92955     petrol        renault          no  
8833    gasoline           audi         yes  
124632    petrol     volkswagen          no  
233625    petrol            bmw         yes  
272265  gasoline          mazda          no  
...          ...            ...         ...  
158838    petrol  mercedes_benz         yes  
47873     petrol     volkswagen          no  
86398   gasoline        peugeot         yes  
77285     petrol     volkswagen          no  
217570  gasoline        renault         yes  

[247320 rows x 9 columns]

In [24]:
# Прямое кодирование для линейных функций 

rev02_features_train_OHE = pd.get_dummies(rev02_features_train, drop_first=True)
rev02_features_test_OHE = pd.get_dummies(rev02_features_test, drop_first=True)
#rev02_target_train_OHE = pd.get_dummies(rev02_target_train, drop_first=True)
#rev02_target_test_OHE = pd.get_dummies(rev02_target_test, drop_first=True)

print(rev02_features_train_OHE.shape, '\n')
display(rev02_features_train_OHE)

#print(rev02_target_train_OHE.shape, '\n')
#display(rev02_target_train_OHE, '\n')

(247320, 307) 



RegistrationYear  Power  Kilometer  VehicleType_convertible  \
92955               1999    107     150000                        0   
8833                2005    140     150000                        0   
124632              2005     75     100000                        0   
233625              1997    184     150000                        0   
272265              2007    143     150000                        0   
...                  ...    ...        ...                      ...   
158838              1995    150     150000                        0   
47873               1991     54     150000                        0   
86398               2011    111      90000                        0   
77285               2000    101     125000                        0   
217570              2006    131     150000                        0   

        VehicleType_coupe  VehicleType_other  VehicleType_sedan  \
92955                   1                  0                  0   
8833                    0                  0                  0   
124632                  0                  0                  1   
233625                  0                  0                  0   
272265                  0                  0                  0   
...                   ...                ...                ...   
158838                  0                  0                  0   
47873                   0                  0                  1   
86398                   0                  0                  0   
77285                   0                  0                  0   
217570                  0                  0                  0   

        VehicleType_small  VehicleType_suv  VehicleType_wagon  ...  \
92955                   0                0                  0  ...   
8833                    1                0                  0  ...   
124632                  0                0                  0  ...   
233625                  0                0                  1  ...   
272265                  0                0                  1  ...   
...                   ...              ...                ...  ...   
158838                  0                0                  1  ...   
47873                   0                0                  0  ...   
86398                   0                0                  1  ...   
77285                   0                0                  1  ...   
217570                  0                0                  1  ...   

        Brand_skoda  Brand_smart  Brand_sonstige_autos  Brand_subaru  \
92955             0            0                     0             0   
8833              0            0                     0             0   
124632            0            0                     0             0   
233625            0            0                     0             0   
272265            0            0                     0             0   
...             ...          ...                   ...           ...   
158838            0            0                     0             0   
47873             0            0                     0             0   
86398             0            0                     0             0   
77285             0            0                     0             0   
217570            0            0                     0             0   

        Brand_suzuki  Brand_toyota  Brand_trabant  Brand_volkswagen  \
92955              0             0              0                 0   
8833               0             0              0                 0   
124632             0             0              0                 1   
233625             0             0              0                 0   
272265             0             0              0                 0   
...              ...           ...            ...               ...   
158838             0             0              0                 0   
47873              0             0              0                 1   
86398

In [25]:
# Порядковое кодирование для "древесных"
encoder = OrdinalEncoder()

rev02_features_train_OE = pd.DataFrame(encoder.fit_transform(rev02_features_train), columns=rev02_features_train.columns)
rev02_features_test_OE = pd.DataFrame(encoder.fit_transform(rev02_features_test), columns=rev02_features_test.columns)
#rev02_target_train_OE = pd.DataFrame(encoder.fit_transform(rev02_target_train), columns=rev02_target_train.columns)
#rev02_target_test_OE = pd.DataFrame(encoder.fit_transform(rev02_target_test), columns=rev02_target_test.columns)

print(rev02_features_train_OE.shape, '\n')
display(rev02_features_train_OE)



(247320, 9) 



VehicleType  RegistrationYear  Gearbox  Power  Model  Kilometer  \
0               2.0              73.0      1.0  106.0  149.0       12.0   
1               5.0              79.0      1.0  139.0   28.0       12.0   
2               4.0              79.0      1.0   74.0  116.0       10.0   
3               7.0              71.0      1.0  183.0   15.0       12.0   
4               7.0              81.0      1.0  142.0   17.0       12.0   
...             ...               ...      ...    ...    ...        ...   
247315          7.0              69.0      0.0  149.0   95.0       12.0   
247316          4.0              65.0      1.0   53.0  116.0       12.0   
247317          7.0              85.0      1.0  110.0   10.0        9.0   
247318          7.0              74.0      1.0  100.0  170.0       11.0   
247319          7.0              80.0      1.0  130.0  149.0       12.0   

        FuelType  Brand  NotRepaired  
0            6.0   27.0          0.0  
1            2.0    1.0          1.0  
2            6.0   38.0          0.0  
3            6.0    2.0          1.0  
4            2.0   19.0          0.0  
...          ...    ...          ...  
247315       6.0   20.0          1.0  
247316       6.0   38.0          0.0  
247317       2.0   25.0          1.0  
247318       6.0   38.0          0.0  
247319       2.0   27.0          1.0  

[247320 rows x 9 columns]

In [26]:
%%time
lmodel = LinearRegression()
lmodel.fit(rev02_features_train_OHE, rev02_target_train)
scores = cross_val_score(lmodel, rev02_features_train_OHE, rev02_target_train, cv=5, scoring='neg_root_mean_squared_error').mean()
print("Linear Regression:", scores) 

Linear Regression: -2783.624817142999
Wall time: 15.2 s


In [27]:
%%time


lmodel = LinearRegression()
lmodel.fit(rev02_features_train_OE, rev02_target_train)

Wall time: 50.3 ms


LinearRegression()

In [28]:
%%time

preds_lmodel = lmodel.predict(rev02_features_train_OE)

Wall time: 4.59 ms


In [29]:
%%time


scores = cross_val_score(lmodel, rev02_features_train_OE, rev02_target_train, cv=5, scoring='neg_root_mean_squared_error').mean()
print("Linear Regression:", scores) 

Linear Regression: -3092.3251219909857
Wall time: 256 ms


In [30]:
from sklearn.tree import DecisionTreeClassifier

### Преобразование модели

In [31]:
# Закодируем категориальные признаки техникой Ordinal Encoding

encoder = OrdinalEncoder()
encoder.fit(end_df) 
data_ordinal = pd.DataFrame(encoder.transform(end_df), columns=end_df.columns)

data_ordinal.reset_index()
end_df.loc[:,['Brand', 'Model']] = data_ordinal.loc[:,['Brand', 'Model']]
end_df['Brand'] = end_df['Brand'].astype(int)
end_df['Model'] = end_df['Model'].astype(int)

Преобразуем категориальные признаки в численные с помощью техники прямого кодирования, или отображения (англ. One-Hot Encoding, OHE)

In [32]:
def dum(end_df, column):
    df = pd.get_dummies(end_df[column], prefix=column, drop_first=True)
    end_df = end_df.drop(column, axis=1)
    return end_df.join(df)

In [33]:
end_df = dum(end_df, 'VehicleType')
end_df = dum(end_df, 'Gearbox')
end_df = dum(end_df, 'FuelType')
end_df = dum(end_df, 'NotRepaired')

In [34]:
#преобразуем все данные через OHE

end_df = pd.get_dummies(end_df, drop_first=True)


In [35]:
end_df.info()
display(end_df.describe())
display(end_df.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 329760 entries, 0 to 329759
Data columns (total 22 columns):
 #   Column                   Non-Null Count   Dtype
---  ------                   --------------   -----
 0   Price                    329760 non-null  int64
 1   RegistrationYear         329760 non-null  int64
 2   Power                    329760 non-null  int64
 3   Model                    329760 non-null  int32
 4   Kilometer                329760 non-null  int64
 5   Brand                    329760 non-null  int32
 6   VehicleType_convertible  329760 non-null  uint8
 7   VehicleType_coupe        329760 non-null  uint8
 8   VehicleType_other        329760 non-null  uint8
 9   VehicleType_sedan        329760 non-null  uint8
 10  VehicleType_small        329760 non-null  uint8
 11  VehicleType_suv          329760 non-null  uint8
 12  VehicleType_wagon        329760 non-null  uint8
 13  Gearbox_manual           329760 non-null  uint8
 14  Gearbox_unknown          329760 non-

Price  RegistrationYear          Power          Model  \
count  329760.000000     329760.000000  329760.000000  329760.000000   
mean     4607.036096       2002.593674     120.046843     114.392722   
std      4546.267728          6.927782      61.371297      73.369348   
min         1.000000       1910.000000       1.000000       0.000000   
25%      1200.000000       1999.000000      75.000000      42.000000   
50%      2900.000000       2003.000000     110.000000     116.000000   
75%      6600.000000       2007.000000     145.000000     170.000000   
max     20000.000000       2016.000000    2000.000000     249.000000   

           Kilometer          Brand  VehicleType_convertible  \
count  329760.000000  329760.000000            329760.000000   
mean   128206.316715      20.807794                 0.060238   
std     37524.363197      13.299661                 0.237927   
min      5000.000000       0.000000                 0.000000   
25%    125000.000000       9.000000                 0.000000   
50%    150000.000000      24.000000                 0.000000   
75%    150000.000000      33.000000                 0.000000   
max    150000.000000      39.000000                 1.000000   

       VehicleType_coupe  VehicleType_other  VehicleType_sedan  ...  \
count      329760.000000      329760.000000      329760.000000  ...   
mean            0.047404           0.069814           0.271109  ...   
std             0.212502           0.254834           0.444533  ...   
min             0.000000           0.000000           0.000000  ...   
25%             0.000000           0.000000           0.000000  ...   
50%             0.000000           0.000000           0.000000  ...   
75%             0.000000           0.000000           1.000000  ...   
max             1.000000           1.000000           1.000000  ...   

       VehicleType_wagon  Gearbox_manual  Gearbox_unknown  FuelType_electric  \
count      329760.000000   329760.000000    329760.000000      329760.000000   
mean            0.193156        0.761636         0.047313           0.000264   
std             0.394775        0.426084         0.212308           0.016241   
min             0.000000        0.000000         0.000000           0.000000   
25%             0.000000        1.000000         0.000000           0.000000   
50%             0.000000        1.000000         0.000000           0.000000   
75%             0.000000        1.000000         0.000000           0.000000   
max             1.000000        1.000000         1.000000           1.000000   

       FuelType_gasoline  FuelType_hybrid   FuelType_lpg  FuelType_other  \
count      329760.000000    329760.000000  329760.000000   329760.000000   
mean            0.287615         0.000667       0.015232        0.073808   
std             0.452651         0.025821       0.122476        0.261459   
min             0.000000         0.000000       0.000000        0.000000   
25%             0.000000         0.000000       0.000000        0.000000   
50%             0.000000         0.000000       0.000000        0.000000   
75%             1.000000         0.000000       0.000000        0.000000   
max             1.000000         1.000000       1.000000        1.000000   

       FuelType_petrol  NotRepaired_yes  
count    329760.000000    329760.000000  
mean          0.620797         0.281559  
std           0.485189         0.449760  
min           0.000000         0.000000  
25%           0.000000         0.000000  
50%           1.000000         0.000000  
75%           1.000000         1.000000  
max           1.000000         1.000000  

[8 rows x 22 columns]

Price  RegistrationYear  Power  Model  Kilometer  Brand  \
0    480              1993    101    116     150000     38   
1  18300              2011    190    227     125000      1   
2   9800              2004    163    117     125000     14   
3   1500              2001     75    116     150000     38   
4   3600              2008     69    101      90000     31   

   VehicleType_convertible  VehicleType_coupe  VehicleType_other  \
0                        0                  0                  1   
1                        0                  1                  0   
2                        0                  0                  0   
3                        0                  0                  0   
4                        0                  0                  0   

   VehicleType_sedan  ...  VehicleType_wagon  Gearbox_manual  Gearbox_unknown  \
0                  0  ...                  0               1                0   
1                  0  ...                  0               1                0   
2                  0  ...                  0               0                0   
3                  0  ...                  0               1                0   
4                  0  ...                  0               1                0   

   FuelType_electric  FuelType_gasoline  FuelType_hybrid  FuelType_lpg  \
0                  0                  0                0             0   
1                  0                  1                0             0   
2                  0                  1                0             0   
3                  0                  0                0             0   
4                  0                  1                0             0   

   FuelType_other  FuelType_petrol  NotRepaired_yes  
0               0                1                1  
1               0                0                1  
2               0                0                1  
3               0                1                0  
4               0                0                0  

[5 rows x 22 columns]

### Подготовка обучающей выборки

In [36]:
# Подготовка тестовой и валидационной выборки, на признаки и целевые признаки согласно заданию
target = end_df['Price']
features = end_df.drop('Price', axis=1)
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.25, random_state=12345)

In [37]:
# глянем, что получилось
features_train

RegistrationYear  Power  Model  Kilometer  Brand  \
92955               1999    107    149     150000     27   
8833                2005    140     28     150000      1   
124632              2005     75    116     100000     38   
233625              1997    184     15     150000      2   
272265              2007    143     17     150000     19   
...                  ...    ...    ...        ...    ...   
158838              1995    150     95     150000     20   
47873               1991     54    116     150000     38   
86398               2011    111     10      90000     25   
77285               2000    101    170     125000     38   
217570              2006    131    149     150000     27   

        VehicleType_convertible  VehicleType_coupe  VehicleType_other  \
92955                         0                  1                  0   
8833                          0                  0                  0   
124632                        0                  0                  0   
233625                        0                  0                  0   
272265                        0                  0                  0   
...                         ...                ...                ...   
158838                        0                  0                  0   
47873                         0                  0                  0   
86398                         0                  0                  0   
77285                         0                  0                  0   
217570                        0                  0                  0   

        VehicleType_sedan  VehicleType_small  ...  VehicleType_wagon  \
92955                   0                  0  ...                  0   
8833                    0                  1  ...                  0   
124632                  1                  0  ...                  0   
233625                  0                  0  ...                  1   
272265                  0                  0  ...                  1   
...                   ...                ...  ...                ...   
158838                  0                  0  ...                  1   
47873                   1                  0  ...                  0   
86398                   0                  0  ...                  1   
77285                   0                  0  ...                  1   
217570                  0                  0  ...                  1   

        Gearbox_manual  Gearbox_unknown  FuelType_electric  FuelType_gasoline  \
92955                1                0                  0                  0   
8833                 1                0                  0                  1   
124632               1                0                  0                  0   
233625               1                0                  0                  0   
272265               1                0                  0                  1   
...                ...              ...                ...                ...   
158838               0                0                  0                  0   
47873                1                0                  0                  0   
86398                1                0                  0                  1   
77285                1                0                  0                  0   
217570               1                0                  0                  1   

        FuelType_hybrid  FuelType_lpg  FuelType_other  FuelType_petrol  \
92955                 0             0               0                1   
8833                  0             0               0                0   
124632                0             0               0                1   
233625                0             0               0                1   
272265                0             0               0                0   
...                 ...           ...             ...              ...   
158838                0             0               0 

In [38]:
# далее обработаем ранее численные признаки
# Для удобства выделим их
numeric = ['RegistrationYear', 'Power', 'Kilometer'] # 'Model', 'Brand'

# Совместим кодирование категориальных и масштабирование численных признаков
scaler = StandardScaler()
scaler.fit(features_train[numeric])
features_train[numeric] = scaler.transform(features_train[numeric])
features_test[numeric] = scaler.transform(features_test[numeric]) 


display(features_train.head())


RegistrationYear     Power  Model  Kilometer  Brand  \
92955          -0.517738 -0.212969    149   0.581307     27   
8833            0.347339  0.324509     28   0.581307      1   
124632          0.347339 -0.734159    116  -0.749625     38   
233625         -0.806097  1.041145     15   0.581307      2   
272265          0.635698  0.373370     17   0.581307     19   

        VehicleType_convertible  VehicleType_coupe  VehicleType_other  \
92955                         0                  1                  0   
8833                          0                  0                  0   
124632                        0                  0                  0   
233625                        0                  0                  0   
272265                        0                  0                  0   

        VehicleType_sedan  VehicleType_small  ...  VehicleType_wagon  \
92955                   0                  0  ...                  0   
8833                    0                  1  ...                  0   
124632                  1                  0  ...                  0   
233625                  0                  0  ...                  1   
272265                  0                  0  ...                  1   

        Gearbox_manual  Gearbox_unknown  FuelType_electric  FuelType_gasoline  \
92955                1                0                  0                  0   
8833                 1                0                  0                  1   
124632               1                0                  0                  0   
233625               1                0                  0                  0   
272265               1                0                  0                  1   

        FuelType_hybrid  FuelType_lpg  FuelType_other  FuelType_petrol  \
92955                 0             0               0                1   
8833                  0             0               0                0   
124632                0             0               0                1   
233625                0             0               0                1   
272265                0             0               0                0   

        NotRepaired_yes  
92955                 0  
8833                  1  
124632                0  
233625                1  
272265                0  

[5 rows x 21 columns]

In [39]:
catfeatures_train = features_train
cattarget_train = target_train
catfeatures_train

RegistrationYear     Power  Model  Kilometer  Brand  \
92955          -0.517738 -0.212969    149   0.581307     27   
8833            0.347339  0.324509     28   0.581307      1   
124632          0.347339 -0.734159    116  -0.749625     38   
233625         -0.806097  1.041145     15   0.581307      2   
272265          0.635698  0.373370     17   0.581307     19   
...                  ...       ...    ...        ...    ...   
158838         -1.094456  0.487380     95   0.581307     20   
47873          -1.671173 -1.076190    116   0.581307     38   
86398           1.212415 -0.147820     10  -1.015811     25   
77285          -0.373558 -0.310692    170  -0.084159     38   
217570          0.491518  0.177924    149   0.581307     27   

        VehicleType_convertible  VehicleType_coupe  VehicleType_other  \
92955                         0                  1                  0   
8833                          0                  0                  0   
124632                        0                  0                  0   
233625                        0                  0                  0   
272265                        0                  0                  0   
...                         ...                ...                ...   
158838                        0                  0                  0   
47873                         0                  0                  0   
86398                         0                  0                  0   
77285                         0                  0                  0   
217570                        0                  0                  0   

        VehicleType_sedan  VehicleType_small  ...  VehicleType_wagon  \
92955                   0                  0  ...                  0   
8833                    0                  1  ...                  0   
124632                  1                  0  ...                  0   
233625                  0                  0  ...                  1   
272265                  0                  0  ...                  1   
...                   ...                ...  ...                ...   
158838                  0                  0  ...                  1   
47873                   1                  0  ...                  0   
86398                   0                  0  ...                  1   
77285                   0                  0  ...                  1   
217570                  0                  0  ...                  1   

        Gearbox_manual  Gearbox_unknown  FuelType_electric  FuelType_gasoline  \
92955                1                0                  0                  0   
8833                 1                0                  0                  1   
124632               1                0                  0                  0   
233625               1                0                  0                  0   
272265               1                0                  0                  1   
...                ...              ...                ...                ...   
158838               0                0                  0                  0   
47873                1                0                  0                  0   
86398                1                0                  0                  1   
77285                1                0                  0                  0   
217570               1                0                  0                  1   

        FuelType_hybrid  FuelType_lpg  FuelType_other  FuelType_petrol  \
92955                 0             0               0                1   
8833                  0             0               0                0   
124632                0             0               0                1   
233625                0             0               0                1   
272265                0             0               0                0   
...                 ...           ...             ...              ...   
158838            

### Обучение моделей

#### Linear Regression

In [40]:
target_train = rev02_target_train
features_train = rev02_features_train_OE
target_test = rev02_target_test
features_test = rev02_features_test_OE

In [41]:
%%time
# время обучения = lmodel.fit(features_train, target_train)

lmodel = LinearRegression()
lmodel.fit(features_train, target_train)


Wall time: 63.1 ms


LinearRegression()

In [42]:
%%time

# время предсказания = lmodel.predict(features_train)
preds_lmodel = lmodel.predict(features_train)

Wall time: 8.48 ms


In [43]:
%%time
# здесь метрику нужно считать по кросс-валидации (cross_val_score()) на тренировочной выборке


#mse_lmodel = mean_squared_error(target_train, preds_lmodel)
scores = cross_val_score(lmodel, features_train, target_train, cv=5, scoring='neg_root_mean_squared_error').mean()
#final_score = sum(scores) / len(scores)
print("Linear Regression:", scores) #round((mse_lmodel) ** 0.5, 2))

Linear Regression: -3092.3251219909857
Wall time: 269 ms


#### Ridge

In [44]:
%%time

ridge_model = Ridge()
hyperparams = [{'solver':['auto', 'svd', 'cholesky', 'lsqr','sparse_cg']}]
ridge_model = GridSearchCV(ridge_model, hyperparams, scoring='neg_mean_squared_error')
ridge_model.fit(features_train, target_train)

print(ridge_model.best_params_)

{'solver': 'auto'}
Wall time: 1.52 s


In [45]:
%%time
preds_ridge_model = ridge_model.predict(features_test)

Wall time: 8.95 ms


In [46]:
mse_ridge_model = mean_squared_error(target_test, preds_ridge_model)
print("RMSE - Ridge:", round((mse_ridge_model) ** 0.5, 2))

RMSE - Ridge: 3745.35


#### DecisionTreeRegresso

In [47]:
%%time

dtr_model = DecisionTreeRegressor(criterion='mse', 
                              max_depth=8, 
                              random_state=12345) 
dtr_model.fit(features_train, target_train)

C:\Users\Bazunov_Anton\anaconda3\lib\site-packages\sklearn\tree\_classes.py:359: FutureWarning: Criterion 'mse' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='squared_error'` which is equivalent.
  warnings.warn(


Wall time: 366 ms


DecisionTreeRegressor(criterion='mse', max_depth=8, random_state=12345)

In [48]:
%%time
preds_dtr_model = dtr_model.predict(features_train)

Wall time: 24 ms


In [49]:
mse_dtr_model = mean_squared_error(target_train, preds_dtr_model)
print("RMSE - DecisionTreeRegresso:", round((mse_dtr_model) ** 0.5, 2))

RMSE - DecisionTreeRegresso: 2110.1


#### CatBoostRegressor

In [50]:
numerical_features = ['Price', 'RegistrationYear', 'Power', 'Kilometer']

In [51]:
categorical_features = [col for col in list(catfeatures_train.columns) if col not in numerical_features]

    %%time
    catmodel = CatBoostRegressor(learning_rate=0.5, random_state=12345, verbose=False) 
    catmodel.fit(features_train, target_train, cat_features=categorical_features)

При попытке обработать в данном формате вылезает ошибка: 
    `"Invalid type for cat_feature[non-default value idx=0,feature_idx=2]=-1.1706373756567139 : cat_features must be integer or string, real number values and NaN values should be converted to string."`

Связано с тем, что CatBoostRegressor принимает, только целые числа или строки в категориях

In [52]:
(df_test.
 loc[:,['VehicleType','Gearbox', 'Brand', 'FuelType', 'NotRepaired']]) = (data_ordinal.
                                                                                loc[:,['VehicleType','Gearbox', 'Brand', 'FuelType', 'NotRepaired']])
catfeatures = df_test.drop('Price', axis=1)
cattarget = df_test['Price']
(catfeatures_train, 
 catfeatures_test, 
 cattarget_train, 
 cattarget_test) = (train_test_split
                    (features, target, test_size=0.25, random_state=12345))

In [53]:
numeric_cat = ['RegistrationYear', 'Power', 'Kilometer']
scaler = StandardScaler()
scaler.fit(catfeatures_train[numeric_cat])
catfeatures_train[numeric_cat] = scaler.transform(catfeatures_train[numeric_cat])
catfeatures_test[numeric_cat] = scaler.transform(catfeatures_test[numeric_cat]) 

In [54]:
%%time
cat_model = CatBoostRegressor(learning_rate=0.5, random_state=12345, verbose=False) 
cat_model.fit(catfeatures_train, cattarget_train, cat_features=categorical_features)

Wall time: 1min 26s


In [55]:
%%time
cat_predict = cat_model.predict(catfeatures_train)

Wall time: 1.33 s


In [56]:
mse_cat_model = mean_squared_error(cattarget_train, cat_predict)
print("RMSE - CatBoostRegressor:", round((mse_cat_model) ** 0.5, 2))

RMSE - CatBoostRegressor: 1445.49


#### Случайный Лес

In [57]:
%%time

model_dt= DecisionTreeRegressor(random_state=12345)
param_grid = {
    "max_depth" : range(15,16),
    "min_samples_leaf" : range(10,11)
}
gsearch = GridSearchCV(model_dt, param_grid = param_grid, scoring='neg_root_mean_squared_error', cv=2)
dt_model = gsearch.fit(features_train,target_train)
print(f'Лучшая RMSE: {-dt_model.best_score_} при параметрах {dt_model.best_params_}')
print('\n', 'Время выполнения:')

Лучшая RMSE: 1913.8570998325458 при параметрах {'max_depth': 15, 'min_samples_leaf': 10}

 Время выполнения:
Wall time: 1.2 s


#### LGBMRegressor

In [58]:
%%time

lgbm_model = LGBMRegressor(random_state=12345, n_jobs=2)
param_grid = {
    "max_depth" : range(10,15),
    "n_estimators" : range(500,505)
}

gsearch = GridSearchCV(lgbm_model, param_grid = param_grid, scoring='neg_root_mean_squared_error', cv=2)
lgb_model = gsearch.fit(features_train, target_train)

print(f'Лучшая RMSE: {-lgb_model.best_score_} при параметрах ,{lgb_model.best_params_}')
print('\n', 'Время выполнения:')

Лучшая RMSE: 1650.0676408861118 при параметрах ,{'max_depth': 13, 'n_estimators': 504}

 Время выполнения:
Wall time: 2min 25s


**Прмоежуточный итог:**
* подготовленны данные для дальнейшего обучения
* обработаны и проверенны ряд моделей

## Анализ моделей

* *Проанализируйте время обучения, время предсказания и качество моделей.*
* *Опираясь на критерии заказчика, выберете лучшую модель, проверьте её качество на тестовой выборке.*

### Объединение и анализ полученных результатов


In [59]:
am_df = [["378 ms", "93.8 ms", 3061.46],
        ["1.72 s", "4.35 ms", 3745.35],
        ["750 ms", "54.1 ms", 2110.10],
        ["2min 7s", "2.34 s", 1445.49],
        ["2.13 s", "-", 1913.85],
        ["2min 25s", "-", 1650.00]]
model = ["Linear Regression", "Ridge", "DecisionTreeRegresso", "CatBoostRegressor", "DecisionTreeRegresso + GridSearchCV", "LGBMRegressor"]

In [60]:
pd.DataFrame(data=am_df, index=model, columns=["Wall_time", "predict_time", "RMSE"])

Wall_time predict_time     RMSE
Linear Regression                      378 ms      93.8 ms  3061.46
Ridge                                  1.72 s      4.35 ms  3745.35
DecisionTreeRegresso                   750 ms      54.1 ms  2106.53
CatBoostRegressor                     2min 7s       2.34 s  1445.49
DecisionTreeRegresso + GridSearchCV    2.13 s            -  1902.47
LGBMRegressor                        6min 29s            -  1650.00

Наилучший результат по RMSE показала модель `CatBoostRegressor` = 1445.49, время выполнения - 1min 26s, на втором месте `LGBMRegressor` = 1650.00 и существенно дольше выполняется - 6min 29s 

Результат отличается не сильно, а время практически в два раза, по этой причине модель `CatBoostRegressor`.

### Проверка финальной метрики

In [62]:
%%time
cat_predict = cat_model.predict(catfeatures_test)

mse_cat_model = mean_squared_error(cattarget_test, cat_predict)
print("Итоговое RMSE - CatBoostRegressor:", round((mse_cat_model) ** 0.5, 2))
print('\n', 'Время выполнения:')


Итоговое RMSE - CatBoostRegressor: 1617.45

 Время выполнения:
Wall time: 499 ms


## Вывод



Результаты финальной метрики удовлетворяют требованиям заказчика => модель подобранна верно.